web_client/views/imageViewerWidget/geojs.js

```diff
@@ -151,11 +154,89 @@ var GeojsImageViewerWidget = ImageViewerWidget.extend({
                         ],
                         (evt) => this._onMouseFeature(evt)
                     );
+
+                    // store the original opacities for the elements in each feature
+                    const data = feature.data();
+                    if (data.length <= this._highlightFeatureSizeLimit) {
+                        this._featureOpacity[annotation.id][feature.featureType] = feature.data()
+                            .map(({id, properties}) => {
+                                return {
+                                    id,
+                                    fillOpacity: properties.fillOpacity,
+                                    strokeOpacity: properties.strokeOpacity
+                                };
+                            });
+                    }
                 });
+                this._mutateFeaturePropertiesForHighlight(annotation.id, features);
                 this.viewer.draw();
             });
     },
 
+    /**
+     * Highlight the given annotation/element by reducing the opacity of all
+     * other elements by 75%.  For performance reasons, features with a large
+     * number of elements are not modified.  The limit for this behavior is
+     * configurable via the constructor option `highlightFeatureSizeLimit`.
+     *
+     * Both arguments are optional.  If no element is provided, then all
+     * elements in the given annotation are highlighted.  If no annotation
+     * is provided, then highlighting state is reset and the original
+     * opacities are used for all elements.
+     *
+     * @param {string?} annotation The id of the annotation to highlight
+     * @param {string?} element The id of the element to highlight
+     */
+    highlightAnnotation: function (annotation, element) {
+        this._highlightAnnotation = annotation;
+        this._highlightElement = element;
+        _.each(this._annotations, (layer, annotationId) => {
+            const features = layer.features;
+            this._mutateFeaturePropertiesForHighlight(annotationId, features);
+        });
+        this.viewer.draw();
+        return this;
+    },
+
+    /**
+     * Use geojs's `updateStyleFromArray` to modify the opacities of all elements
+     * in a feature.  This method uses the private attributes `_highlightAnntotation`
+     * and `_highlightElement` to determine which element to modify.
+     */
+    _mutateFeaturePropertiesForHighlight: function (annotationId, features) {
+        _.each(features, (feature) => {
+            const data = this._featureOpacity[annotationId][feature.featureType];
```

Comment:
Is there ever a case where this could be called before `drawAnnotation`?  If so, there should be a guard to ensure that `this._featureOpacity[annotationId]` exists.

Reply:
No, not directly.  From the call [here](https://github.com/girder/large_image/blob/5273ae00bab6ab0cf84896dbef87893ae95a6b93/web_client/views/imageViewerWidget/geojs.js#L195), it is only called if `this._annotations[id]` exists which is created synchronously with the `this._featureOpacity[id]` object in `drawAnnotation`.